apps/src/code-studio/initApp/project.js

```diff
@@ -1923,6 +1923,22 @@ var projects = (module.exports = {
     }
     return sourcesApi;
   },
+
+  // Sets a callback to save the project before unloading the page.
+  registerSaveOnUnload() {
+    const unloadHandler = event => {
+      if (this.hasOwnerChangedProject()) {
```

Comment:
Cool! Just tracing through how this works. I see that `hasProjectChanged` is set to `true` when the `events.workspaceChange` is detected.
This event is defined at https://github.com/code-dot-org/code-dot-org/blob/3b049ee796cd976c4235996d9f57e7e3738adb21/apps/src/code-studio/initApp/project.js#L46-L47.
I see where the other events are triggered like `appModeChanged` [Example](https://github.com/code-dot-org/code-dot-org/blob/3b049ee796cd976c4235996d9f57e7e3738adb21/apps/src/StudioApp.js#L2055-L2056)
But could you point out where the `events.workspaceChange` is actually triggered? Just trying to understand how this is working. Thanks!

Reply:
Yeah I was confused by this as well! I just did some digging, and I was able to trigger it on a CDO Blockly level (/s/outbreak/lessons/1/levels/2), but not Google Blockly (I tried dance). It was triggered here:

https://github.com/code-dot-org/blockly/blob/f012d8262f21bae3e54fb11dd8bc29cf0d29f3cd/core/ui/block_space/block_space.js#L831

So, I think that event is only relevant for old Blockly labs.

For Applab, that line I added above (`studioApp().addChangeHandler(project.projectChanged);`) is what triggers the change to the `hasProjectChanged` state:

https://github.com/code-dot-org/code-dot-org/blob/41dc24fe1c3a371469ca95e1e46326a57120b1cf/apps/src/StudioApp.js#L1044-L1075